editor/xml_feature.cpp

```diff
@@ -0,0 +1,214 @@
+#include "editor/xml_feature.hpp"
+
+#include "base/assert.hpp"
+#include "base/string_utils.hpp"
+#include "base/timer.hpp"
+
+#include "geometry/mercator.hpp"
+
+#include "std/set.hpp"
+#include "std/unordered_set.hpp"
+
+#include <boost/functional/hash.hpp>
+
+#include "3party/pugixml/src/pugixml.hpp"
+
+namespace
+{
+auto constexpr kLatLonTolerance = 7;
+
+pugi::xml_node FindTag(pugi::xml_document const & document, string const & key)
+{
+  return document.select_node(("//tag[@k='" + key + "']").data()).node();
+}
+
+m2::PointD PointFromLatLon(pugi::xml_node const node)
+{
+  double lat, lon;
+  if (!strings::to_double(node.attribute("lat").value(), lat))
+    MYTHROW(editor::XMLFeatureNoLatLonError,
+            ("Can't parse lat attribute: " + string(node.attribute("lat").value())));
+ if (!strings::to_double(node.attribute("lon").value(), lon))
+   MYTHROW(editor::XMLFeatureNoLatLonError,
+           ("Can't parse lon attribute: " + string(node.attribute("lon").value())));
+  return MercatorBounds::FromLatLon(lat, lon);
+}
+
+void ValidateNode(pugi::xml_node const & node)
+{
+  if (!node)
+    MYTHROW(editor::XMLFeatureNoNodeError, ("Document has no node"));
+
+  try
+  {
+    PointFromLatLon(node);
+  }
+  catch(editor::XMLFeatureError const &)
+  {
+    throw;
+  }
+
+  if (!node.attribute("timestamp"))
+    MYTHROW(editor::XMLFeatureNoTimestampError, ("Node has no timestamp attribute"));
+}
+} // namespace
+
+namespace editor
+{
+char const * const XMLFeature::kLastModified = "timestamp";
+char const * const XMLFeature::kHouseNumber = "addr:housenumber";
+char const * const XMLFeature::kDefaultName = "name";
+char const * const XMLFeature::kLocalName = "name:";
+char const * const XMLFeature::kDefaultLang =
+    StringUtf8Multilang::GetLangByCode(StringUtf8Multilang::DEFAULT_CODE);;
+
+XMLFeature::XMLFeature()
+{
+  m_document.append_child("node");
+}
+
+XMLFeature::XMLFeature(string const & xml)
+{
+  m_document.load(xml.data());
+  ValidateNode(GetRootNode());
+}
+
+XMLFeature::XMLFeature(pugi::xml_document const & xml)
+{
+  m_document.reset(xml);
+  ValidateNode(GetRootNode());
+}
+
+XMLFeature::XMLFeature(pugi::xml_node const & xml)
+{
+  m_document.reset();
+  m_document.append_copy(xml);
+  ValidateNode(GetRootNode());
+}
+
+void XMLFeature::Save(ostream & ost) const
+{
+  m_document.save(ost, "  ", pugi::format_indent_attributes);
+}
+
+m2::PointD XMLFeature::GetCenter() const
```

Comment:
In general, if you have exceptions in constructors, I don't like the idea of throwing getters.